In [8]:
import numpy as np

# 1. 

In [6]:
x = 1
a = 64
inicial = 10
anterior = -1
for i in range(10):
    print('== Interação', i, '==')
    prox = (inicial + (a/inicial)) / 2
    print("x: ",prox)
    
    erro = (abs(prox-anterior)/prox)*100
    print("Erro:", erro)    
    print()
    
    anterior = prox
    inicial = prox

== Interação 0 ==
x:  8.2
Erro: 112.19512195121952

== Interação 1 ==
x:  8.002439024390243
Erro: 2.4687595245351996

== Interação 2 ==
x:  8.000000371689179
Erro: 0.030483157347023407

== Interação 3 ==
x:  8.000000000000009
Erro: 4.6461146263609925e-06

== Interação 4 ==
x:  8.0
Erro: 1.1102230246251565e-13

== Interação 5 ==
x:  8.0
Erro: 0.0

== Interação 6 ==
x:  8.0
Erro: 0.0

== Interação 7 ==
x:  8.0
Erro: 0.0

== Interação 8 ==
x:  8.0
Erro: 0.0

== Interação 9 ==
x:  8.0
Erro: 0.0



# 2.

In [11]:
e = 1
while((1 + e) > 1):
    if 1 + e <= 1:
        e = 2 * e
    else:
        e = e / 2
e = 2*e
e2 = np.finfo(float).eps
print("Epsilon:", e)
print('Epsilon da máquina:', e2)

Epsilon: 2.220446049250313e-16
Epsilon da máquina: 2.220446049250313e-16


In [12]:
if e == e2:
    print("Épsilon da máquinha igual ao épsilon do algoritmo !!")
else:
    print("Épsilon da máquinha diferente ao épsilon do algoritmo !!")

Épsilon da máquinha igual ao épsilon do algoritmo !!


# 3. 

In [14]:
x = 1/3
for i in range(30):
    x = (4 * x) - 1
    print(x)

0.33333333333333326
0.33333333333333304
0.33333333333333215
0.3333333333333286
0.3333333333333144
0.33333333333325754
0.33333333333303017
0.3333333333321207
0.3333333333284827
0.3333333333139308
0.3333333332557231
0.3333333330228925
0.3333333320915699
0.3333333283662796
0.3333333134651184
0.33333325386047363
0.33333301544189453
0.3333320617675781
0.3333282470703125
0.33331298828125
0.333251953125
0.3330078125
0.33203125
0.328125
0.3125
0.25
0.0
-1.0
-5.0
-21.0


Vimos que na primeira interação ele da o valor **0.33333333333333326** ao invés de **0.33333 ...** . Então a cada interação o resultando vai mudando por causa desse valor.  

# 4. 

In [32]:
def calcular(micro):
    print("Micro:", micro)
    print(np.exp(1/micro) / 1 + np.exp(1/micro))
    print(1/np.exp(-1/micro) + 1)
    print()

In [34]:
calcular(0.1)
calcular(0.01)
calcular(0.001)
calcular(0.0001)

Micro: 0.1
44052.931589613436
22027.465794806714

Micro: 0.01
5.376234283632271e+43
2.688117141816135e+43

Micro: 0.001
inf
inf

Micro: 0.0001
inf
inf



<ipython-input-32-6334e7323047>:3: RuntimeWarning: overflow encountered in exp
  print(np.exp(1/micro) / 1 + np.exp(1/micro))
<ipython-input-32-6334e7323047>:4: RuntimeWarning: divide by zero encountered in double_scalars
  print(1/np.exp(-1/micro) + 1)


Quando µ é pequeno, e
1/µ é um número grande. A primeira expressão produz um ”overflow” (número maior que o
máximo representável) quando µ é pequeno. A segunda expressão, no entanto, reproduz o limite 1 quando µ → 0+.

# 5.

In [18]:
def f(x):
    print("x(" + str(x) + ") =", ((1 + x) - 1) / x)

f(10**-12)
f(10**-15)
f(10**-16)
f(10**-17)

x(1e-12) = 1.000088900582341
x(1e-15) = 1.1102230246251565
x(1e-16) = 0.0
x(1e-17) = 0.0


Quando o valor inserido na função é menor que e^15 chega ao máximo que o computador pode calcular de acordo com o épsilon da máquina(float64).